### PR TITLE
[Draft] Try to build with flang 18 on Windows

### PR DESCRIPTION
**Pull request by traversaro**
(do not merge, just an experiment related to https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/1359)

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by traversaro** [comment]
@conda-forge-admin, please rerender

**Comment by traversaro** [comment]
@conda-forge-admin, please rerender

**Comment by traversaro** [comment]
@conda-forge-admin, please rerender

**Comment by traversaro** [comment]
@conda-forge-admin, please rerender

**Comment by traversaro** [comment]
Current failures:

~~~
2024-02-05T12:39:10.7068303Z -- The Fortran compiler identification is LLVMFlang 18.1.0
2024-02-05T12:39:11.6217895Z -- The C compiler identification is Clang 18.1.0 with MSVC-like command-line
2024-02-05T12:39:11.8279108Z -- Detecting Fortran compiler ABI info
2024-02-05T12:39:28.3126243Z -- Detecting Fortran compiler ABI info - done
2024-02-05T12:39:28.8365133Z -- Check for working Fortran compiler: D:/bld/mumps_1707136264872/_build_env/Library/bin/flang-new.exe - skipped
2024-02-05T12:39:28.8366882Z -- Detecting C compiler ABI info
2024-02-05T12:39:29.5516874Z -- Detecting C compiler ABI info - done
2024-02-05T12:39:29.5531558Z -- Check for working C compiler: D:/bld/mumps_1707136264872/_build_env/Library/bin/clang-cl.exe - skipped
2024-02-05T12:39:29.5547018Z -- Detecting C compile features
2024-02-05T12:39:29.5564718Z -- Detecting C compile features - done
2024-02-05T12:39:29.5644183Z -- Looking for Fortran sgemm
2024-02-05T12:39:29.9181059Z -- Looking for Fortran sgemm - not found
2024-02-05T12:39:29.9228196Z -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
2024-02-05T12:39:30.3316649Z -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
2024-02-05T12:39:30.3318355Z -- Looking for pthread_create in pthreads
2024-02-05T12:39:31.1050294Z -- Looking for pthread_create in pthreads - not found
2024-02-05T12:39:31.1051569Z -- Looking for pthread_create in pthread
2024-02-05T12:39:31.2674221Z -- Looking for pthread_create in pthread - not found
2024-02-05T12:39:31.2693903Z -- Found Threads: TRUE  
2024-02-05T12:39:32.4997284Z -- Looking for Fortran sgemm
2024-02-05T12:39:33.3419031Z -- Looking for Fortran sgemm - found
2024-02-05T12:39:33.3427729Z -- Found BLAS: D:/bld/mumps_1707136264872/_h_env/Library/lib/blas.lib  
2024-02-05T12:39:33.6998312Z -- Looking for Fortran cheev
2024-02-05T12:39:34.1690772Z -- Looking for Fortran cheev - not found
2024-02-05T12:39:34.2434860Z -- Looking for Fortran cheev
2024-02-05T12:39:35.0787250Z -- Looking for Fortran cheev - found
2024-02-05T12:39:35.0796885Z -- Found LAPACK: D:/bld/mumps_1707136264872/_h_env/Library/lib/lapack.lib;D:/bld/mumps_1707136264872/_h_env/Library/lib/blas.lib  
2024-02-05T12:39:35.0867781Z -- Configuring done (37.3s)
2024-02-05T12:39:35.3008197Z -- Generating done (0.2s)
2024-02-05T12:39:35.3310955Z -- Build files have been written to: D:/bld/mumps_1707136264872/work/build
2024-02-05T12:39:35.3921296Z ninja: error: '/WHOLEARCHIVE:Fortran_main.static.lib', needed by 'c_example.exe', missing and no known rule to make it
2024-02-05T12:39:37.2138372Z Traceback (most recent call last):
~~~

interestingly, the library installed by https://anaconda.org/conda-forge/libfortran-main is called `Fortran_main.lib`, not `Fortran_main.static.lib` as it seems that CMake is trying to use.

**Comment by h-vetinari** [comment]
> @h-vetinari are you interested in PRs on the rc branch of flang-feedstock, or are you still working on that so you prefer to handle that yourself? Thanks a lot in advance!

Thanks for the analysis, would gladly take a PR! I've noted some relevant changes during the 18 cycle in an issue on the flang feedstock, but I haven't analysed all consequences. This looks like it's one of them.

**Comment by traversaro** [comment]
Actually, I am not sure about what I written, it seems that `Fortran_main.static.lib` is just installed in the `flang` package, while it could make sense to have it in `libfortran-main`, everything should work fine even if it is installed by `flang`.

**Comment by traversaro** [comment]
Reading better the ninja error, I think that in `/WHOLEARCHIVE:Fortran_main.static.lib` the problem is not the library name, but the `/WHOLEARCHIVE:` part.

**Comment by traversaro** [comment]
Manually removing the problematic dependency from build.ninja, the compilation fails with this error:
~~~
[232/912] Building Fortran object CMakeFiles\mumps_common.dir\src\double_linked_list.F.obj
FAILED: CMakeFiles/mumps_common.dir/src/double_linked_list.F.obj mumps_ddll.mod mumps_idll.mod
C:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\_build_env\Library\bin\flang-new.exe -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\src -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\include -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\src -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\PORD\include -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\libseq -ffixed-line-length-72 -o CMakeFiles\mumps_common.dir\src\double_linked_list.F.obj -c CMakeFiles\mumps_common.dir\src\double_linked_list.F-pp.f
error: loc("C:\\Users\\straversaro\\AppData\\Local\\miniforge3\\conda-bld\\mumps_1707139278047\\work\\src\\double_linked_list.F":32:11): operation with block successors must terminate its parent block
error: verification of lowering to FIR failed
[233/912] Building Fortran object CMakeFiles\mumps_common.dir\src\front_data_mgt_m.F.obj
FAILED: CMakeFiles/mumps_common.dir/src/front_data_mgt_m.F.obj mumps_front_data_mgt_m.mod
C:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\_build_env\Library\bin\flang-new.exe -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\src -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\include -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\src -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\PORD\include -IC:\Users\straversaro\AppData\Local\miniforge3\conda-bld\mumps_1707139278047\work\libseq -ffixed-line-length-72 -o CMakeFiles\mumps_common.dir\src\front_data_mgt_m.F.obj -c CMakeFiles\mumps_common.dir\src\front_data_mgt_m.F-pp.f
error: loc("C:\\Users\\straversaro\\AppData\\Local\\miniforge3\\conda-bld\\mumps_1707139278047\\work\\src\\front_data_mgt_m.F":92:7): operation with block successors must terminate its parent block
error: verification of lowering to FIR failed
ninja: build stopped: subcommand failed.
~~~

**Comment by h-vetinari** [comment]
```
error: verification of lowering to FIR failed
```
sounds like flang isn't yet ready for certain syntax constructs (FIR is an intermediate representation during compilation).

Did mumps build with flang 17 or is this the first time you're attempting it? If it's the former, then it's a regression. If it's the latter, it's probably on someone's to-do list already. But in any case, it's probably not a bad idea to raise an issue upstream.

**Comment by Krande** [comment]
Hey @traversaro and @h-vetinari 

Now that the "lowering to FIR" issue in conda-forge `flang` was resolved in https://github.com/conda-forge/flang-feedstock/issues/46 I thought I'd check this PR. 

And it seems like with the fix for flang this PR successfully compiles with only a minor addition.

All I had to do was to add `msvcrt.lib` to `LDFLAGS` in the `bld.seq.bat`

```
set "LDFLAGS=%LDFLAGS% msvcrt.lib"
```

in order to resolve some linking issues

```
[907/913] Linking Fortran shared library zmumps.dll
FAILED: zmumps.dll zmumps.lib
...
mumps_common.lib(mumps_io.c.obj) : error LNK2019: unresolved external symbol __imp_fopen referenced in function mumps_dumprhsbinary_c_
mumps_common.lib(mumps_io_basic.c.obj) : error LNK2001: unresolved external symbol __imp_fopen
mumps_common.lib(mumps_io_err.c.obj) : error LNK2019: unresolved external symbol __imp_strncpy referenced in function mumps_io_error
mumps_common.lib(mumps_save_restore_C.c.obj) : error LNK2001: unresolved external symbol __imp_strncpy
mumps_common.lib(mumps_io_err.c.obj) : error LNK2019: unresolved external symbol __imp_strdup referenced in function mumps_io_sys_error
oldnames.lib(strdup.obi) : error LNK2001: unresolved external symbol __imp_strdup
mumps_common.lib(elapse.c.obj) : error LNK2019: unresolved external symbol __imp__ftime64 referenced in function mumps_elapse_
mumps_common.lib(mumps_io_basic.c.obj) : error LNK2019: unresolved external symbol __imp_fseek referenced in function mumps_io_do_write_block
mumps_common.lib(mumps_io_basic.c.obj) : error LNK2019: unresolved external symbol __imp_ferror referenced in function mumps_io_do_write_block
mumps_common.lib(mumps_io_basic.c.obj) : error LNK2019: unresolved external symbol __imp_fread referenced in function mumps_io_do_read_block
mumps_common.lib(mumps_io_basic.c.obj) : error LNK2019: unresolved external symbol __imp_ftell referenced in function mumps_compute_file_size
oldnames.lib(strdup.obi) : error LNK2001: unresolved external symbol __imp__strdup
zmumps.dll : fatal error LNK1120: 9 unresolved externals
ninja: build stopped: subcommand failed.
```

**Comment by traversaro** [comment]
I wonder if there is some NODEFAULTLIB option being passed by someone? We could also test compiling with cl.exe instead of clang-cl.exe, for consistency with the rest of conda-forge

**Comment by Krande** [comment]
@conda-forge-admin, please rerender

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -9,4 +9,7 @@ pin_run_as_build:
     max_pin: x.x
   scalapack:
     min_pin: x.x.x
-    max_pin: x.x
\ No newline at end of file
+    max_pin: x.x
+
+channel_sources:  # [win]
+  - conda-forge  # [win]
```
Comment (traversaro) [inline recipe/conda_build_config.yaml:15]:
```suggestion
```

This was only added to get flang pre-release, if we do not need it we can remove the complete block.

**Comment by Krande** [comment]
@conda-forge-admin, please rerender

**Comment by Krande** [comment]
@traversaro If we want to build against metis and libscotch. I reckon we need to do something like scivision are doing in their [FindMETIS.cmake](https://github.com/Krande/mumps/blob/main/cmake/FindMETIS.cmake)?

**Comment by traversaro** [comment]
> @traversaro If we want to build against metis and libscotch. I reckon we need to do something like scivision are doing in their [FindMETIS.cmake](https://github.com/Krande/mumps/blob/main/cmake/FindMETIS.cmake)?

Yes, I guess so.

**Comment by h-vetinari** [comment]
> And it seems like with the fix for flang this PR successfully compiles with only a minor addition.
> 
> All I had to do was to add `msvcrt.lib` to `LDFLAGS` in the `bld.seq.bat`
> 
> ```
> set "LDFLAGS=%LDFLAGS% msvcrt.lib"
> ```

That's great news! 🥳

CC also @isuruf

**Comment by h-vetinari** [comment]
> I wonder if there is some NODEFAULTLIB option being passed by someone?

We do set `--dependent-lib=msvcrt` for [clang](https://github.com/conda-forge/clang-win-activation-feedstock/blob/main/recipe/activate-clang_win-64.bat#L13), but looking at this now, I'm pretty sure our flang [activation](https://github.com/conda-forge/flang-feedstock/blob/main/recipe/activate.bat) when used together with msvc is lacking...

File: recipe/meta.yaml [hunk 2]
```diff
@@ -16,14 +16,15 @@ source:
...
+  number: 5
   skip: true  # [win and vc<14]
 
 requirements:
   build:
     # dummy unused top-level requirements
     # needed to construct the right conda-smithy matrix
-    - {{ compiler('fortran') }}
+    - {{ compiler('fortran') }}  # [not win]
+    - flang  # [win]
```
Comment (h-vetinari) [inline recipe/meta.yaml:27]: Note that we already set `fortran_compiler: flang` on windows, so you can leave the compiler-jinja also on windows, and simply set 
```
fortran_compiler_version:  # [win]
  - 18                     # [win]
```
in CBC

**Comment by h-vetinari** [comment]
> We do set `--dependent-lib=msvcrt` for [clang](https://github.com/conda-forge/clang-win-activation-feedstock/blob/main/recipe/activate-clang_win-64.bat#L13) [...]

Note that flang only gained support for this option in LLVM 18, see https://github.com/conda-forge/flang-feedstock/issues/35

File: recipe/meta.yaml (hunk 2 above)
Comment (Krande) [reply recipe/meta.yaml]: Done.

**Comment by Krande** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml (hunk 2 above)
Comment (h-vetinari) [reply recipe/meta.yaml]: The specification of `fortran_compiler` is [redundant](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/74c2e63226d566984c6219aaf8b6629ea0b48bf9/recipe/conda_build_config.yaml#L38) with the global pinning, but 👍

**Comment by h-vetinari** [comment]
I rebased this PR in #119 for testing with the upcoming flang 19

**Comment by traversaro** [comment]
> I rebased this PR in #119 for testing with the upcoming flang 19

Great! Let's close this PR as it was superseded by https://github.com/conda-forge/mumps-feedstock/pull/119 .

**Pull request closed without merging**